src/validation.cpp

```diff
@@ -1278,11 +1278,9 @@ void UpdateCoins(const CTransaction& tx, CCoinsViewCache& inputs, CTxUndo &txund
             // mark an outpoint spent, and construct undo information
             txundo.vprevout.push_back(CTxInUndo(coins->vout[nPos]));
             coins->Spend(nPos);
-            if (coins->vout.size() == 0) {
-                CTxInUndo& undo = txundo.vprevout.back();
-                undo.nHeight = coins->nHeight;
-                undo.fCoinBase = coins->fCoinBase;
-            }
+            CTxInUndo& undo = txundo.vprevout.back();
+            undo.nHeight = coins->nHeight;
+            undo.fCoinBase = coins->fCoinBase;
```

Comment:
> Will older software just log errors and keep working, or require re-indexing, or will it not work at all?

Older code will fail during the start-up consistency check (the rollback test will notice entries with undo data present where it isn't expected), and fail to start.

> But I also thought this was one of the cases that DisconnectBlock would "graciously deal with"

Well, it does deal with it correctly (it's DISCONNECT_UNCLEAN case, not a DISCONNECT_FAILED one), but the code calling DisconnectBlock requires a DISCONNECT_OK.

In #10148, DisconnectBlock is called to rewind blocks for replay at startup, and there DISCONNECT_UNCLEAN is acceptable.